what4/src/What4/Serialize/Normalize.hs

```diff
@@ -0,0 +1,158 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE MultiWayIf #-}
+
+-- | Normalization and equivalence checking for expressions
+module What4.Serialize.Normalize
+  ( normSymFn
+  , normExpr
+  , testEquivSymFn
+  , testEquivExpr
+  , ExprEquivResult(..)
+  ) where
+
+import qualified Data.Parameterized.Context as Ctx
+import qualified Data.Parameterized.TraversableFC as FC
+
+import qualified What4.Interface as S
+import qualified What4.Expr as S
+import qualified What4.Expr.Builder as B
+import qualified What4.Expr.WeightedSum as WSum
+import           Data.Parameterized.Classes
+
+
+normSymFn :: forall sym st fs t args ret. sym ~ B.ExprBuilder t st fs
```

Comment:
What exactly is being normalized in this function? Worth commenting.

Reply:
I added some commentary on what I think this is doing, but I also am not entirely sure what the full set of normalizations is.